parlai/core/metrics.py

```diff
@@ -115,39 +147,157 @@ class AverageMetric(Metric):
     Class that keeps a running average of some metric.
     """
 
-    def __init__(self, numer: TScalar, denom: TScalar):
-        self._numer = self.as_float(numer)
-        self._denom = self.as_int(denom)
+    __slots__ = ('_numer', '_denom')
+
+    def __init__(self, numer: TScalar, denom: TScalar = 1):
+        self._numer = self.as_number(numer)
+        self._denom = self.as_number(denom)
 
-    def __add__(self, other: 'AverageMetric') -> 'AverageMetric':
+    def __add__(self, other: Optional['AverageMetric']) -> 'AverageMetric':
         # NOTE: hinting can be cleaned up with "from __future__ import annotations" when
         # we drop Python 3.6
-        full_numer: float = self._numer + other._numer
-        full_denom: int = self._denom + other._denom
+        if other is None:
+            return self
+        full_numer: TScalar = self._numer + other._numer
+        full_denom: TScalar = self._denom + other._denom
         return AverageMetric(numer=full_numer, denom=full_denom)
 
     def value(self) -> float:
         return self._numer / self._denom
 
 
-def normalize_answer(s):
+class F1Metric(AverageMetric):
     """
-    Lower text and remove punctuation, articles and extra whitespace.
+    Helper class which computes token-level F1.
     """
 
-    def remove_articles(text):
-        return re_art.sub(' ', text)
+    @staticmethod
+    def _prec_recall_f1_score(pred_items, gold_items):
+        """
+        Compute precision, recall and f1 given a set of gold and prediction items.
+
+        :param pred_items: iterable of predicted values
+        :param gold_items: iterable of gold values
+
+        :return: tuple (p, r, f1) for precision, recall, f1
+        """
+        common = Counter(gold_items) & Counter(pred_items)
+        num_same = sum(common.values())
+        if num_same == 0:
+            return 0, 0, 0
+        precision = 1.0 * num_same / len(pred_items)
+        recall = 1.0 * num_same / len(gold_items)
+        f1 = (2 * precision * recall) / (precision + recall)
+        return precision, recall, f1
+
+    @staticmethod
+    def compute(guess: str, answers: List[str]) -> 'F1Metric':
+        if guess is None or answers is None:
+            return AverageMetric(0, 0)
+        g_tokens = normalize_answer(guess).split()
```

Comment:
stray thought: it seems we always split on spaces after normalizing the answer. can we just make `normalize_answer` return the split list?